catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/forms/model/JsonTransformVisitor.java

```diff
@@ -0,0 +1,238 @@
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.ui.forms.model;
+
+import com.google.common.collect.ImmutableMap;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.xml.bind.JAXBElement;
+import net.opengis.filter.v_2_0.BinaryComparisonOpType;
+import net.opengis.filter.v_2_0.BinaryLogicOpType;
+import net.opengis.filter.v_2_0.BinarySpatialOpType;
+import net.opengis.filter.v_2_0.DistanceBufferType;
+import net.opengis.filter.v_2_0.FunctionType;
+import net.opengis.filter.v_2_0.LiteralType;
+import org.codice.ddf.catalog.ui.forms.filter.AbstractFilterVisitor2;
+import org.codice.ddf.catalog.ui.forms.filter.FilterProcessingException;
+import org.codice.ddf.catalog.ui.forms.filter.VisitableXmlElement;
+import org.codice.ddf.catalog.ui.forms.model.pojo.FilterNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Traverses a Filter represented by some root {@link VisitableXmlElement} and builds a
+ * UI-compatible {@link FilterNode} that can be serialized as part of a model. This allows the
+ * persistence layer to contain standards-compliant Filter XML 2.0 which is transformed on-the-fly
+ * for the UI to consume.
+ *
+ * <p>Irrespective of errors in the bindings, at the moment, this visitor makes an assumption: Value
+ * references support a {@link List<Serializable>} but for our purposes we are only interested in
+ * the first value. The rest are ignored.
+ *
+ * <p>To get started, load XML using JAXB, create a {@link VisitableXmlElement} from the root {@link
+ * JAXBElement}, and pass an instance of {@link JsonTransformVisitor} into {@link
+ * VisitableXmlElement#accept(org.codice.ddf.catalog.ui.forms.filter.FilterVisitor2)}.
+ *
+ * <p>While trying to wrap one's head around visitors, it would be helpful to set a breakpoint on
+ * {@link #visitLiteralType(VisitableXmlElement)} and take some time analyzing the stack.
+ *
+ * <p><i>This code is experimental. While it is functional and tested, it may change or be removed
+ * in a future version of the library.</i>
+ */
+public class JsonTransformVisitor extends AbstractFilterVisitor2 {
+  private static final Logger LOGGER = LoggerFactory.getLogger(JsonTransformVisitor.class);
+
+  private static final Function<Serializable, Boolean> BOOL_FUNC =
+      s -> Boolean.parseBoolean((String) s);
+
+  private static final String TEMPLATE_SUBSTITUTIONS_V1 = "template.substitutions.1";
```

Comment:
❗️ This only represents a single substitution so it should not be plural.  It is also only used to substitute values in the filter.  There is also little context of what the number represents in the templates. Maybe `template.value.v1` or `substitution.value.v1`.